### ISSUE TITLE
document.createRange isn't polyfilled in jsdom, causes exception in testing

### ISSUE THREAD

**Issue by frontsideair** [FezVrasta/popper.js#478]
### CodePen demo

https://codepen.io/pen?template=wGqJEz

### Steps to reproduce the problem

1.  Run UI tests with jsdom
2.  Get exception `TypeError: document.createRange is not a function`

### What is the expected behavior?

Not to get an exception

### What went wrong?

Obviously this is not popper's fault, but maybe it can be fixed here.

### Any other comments?

Thanks for the awesome library, I hope this doesn't come across as disrespectful. The [jsdom issue](https://github.com/tmpvar/jsdom/issues/317) gives the feeling that this won't implemented anytime soon. I would like to know more about why `createRange` is used and can we replace it with something else or fall back to another solution if the browser doesn't support it.

Browsers newer than IE8 already [support](https://caniuse.com/#feat=dom-range) this so I can't find another use case other than testing with `jsdom`.

**Comment by FezVrasta** [comment]
I decided to use createRange because it's a convenient way to know the order of two given DOM elements in the DOM.

Previously I used a very complex and unreliable custom code that didn't perform so well.

I may suggest you to simply polyfill `createRange` since anyway Popper.js is not going to actually work inside JSDOM.

**Comment by frontsideair** [comment]
This turned out to be trickier than I imagined. I polyfilled `document.createRange` but now I get `NaN` for some computed styles, the reason is `jsdom` returns `0` for every property of the bounding client rect.

I imagine this will never be the case in a real browser, but if there's a possibility if it can be, then guarding against NaN (or more generally non-finite values) could be an improvement.

**Comment by FezVrasta** [comment]
Yes as I said, Popper.js is probably never going to work in JSDOM, it relies on real browser DOM APIs to work.

**Comment by frontsideair** [comment]
Of course, I don't expect it to work in a simulated environment. I just want it to not cause errors in snapshot tests. `jsdom` is a big and old project with high barrier of entry for me, but I'm willing to work on Popper.js to fix this issue. 😺

**Comment by FezVrasta** [comment]
Wouldn't make sense to simply mock Popper.js?

**Comment by brentertz** [comment]
@frontsideair I am dealing with the exact same thing.  If you come up with a solution, please share, and I'll do the same.

**Comment by FezVrasta** [comment]
Since on JSDom the only thing that Popper.js will do is to add some wrong inline style to the popper element, why not mock it away?

```js
jest.mock('popper.js', () => class {
  constructor() { return {
    scheduleUpdate: jest.fn(),
    update: jest.fn(),
  };
});
```

or something like that?

**Comment by brentertz** [comment]
Thanks @FezVrasta The following works for me.

```js
jest.mock(
  'popper.js',
  () =>
    class Popper {
      static placements = [
        'auto',
        'auto-end',
        'auto-start',
        'bottom',
        'bottom-end',
        'bottom-start',
        'left',
        'left-end',
        'left-start',
        'right',
        'right-end',
        'right-start',
        'top',
        'top-end',
        'top-start'
      ];

      constructor() {
        return {
          destroy: () => {},
          scheduleUpdate: () => {}
        };
      }
    }
);
```

**Comment by FezVrasta** [comment]
You can also use `static placements = Popper.placements` I think, or anyway there's a way to access the original module so that you don't have to copy the placements.

**Comment by brentertz** [comment]
Yeah, I am experimenting with that now, but currently Jest is complaining about it.

**Comment by FezVrasta** [comment]
I remember there was a syntax that looked like:

```
jest.mock(originalImplementation => {
  =>
    class Popper {
      static placements = originalImplementation.placements;

      constructor() {
        return {
          destroy: () => {},
          scheduleUpdate: () => {}
        };
      }
    }
);
```

but I can't find it now

**Comment by brentertz** [comment]
Boom 💥 

```js
jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {}
      };
    }
  };
});
```

**Comment by FezVrasta** [comment]
That's it! Thanks :-)

**Comment by brentertz** [comment]
Just to close the loop on this...

I also discovered that following works and may be more desirable for some folks.

Create a folder/file called `__mocks__/popper.js.js` adjacent to your `node_modules` directory containing the following code.  Jest will pick it up automatically.

```js
import PopperJs from 'popper.js';

export default class Popper {
  static placements = PopperJs.placements;

  constructor() {
    return {
      destroy: () => {},
      scheduleUpdate: () => {}
    };
  }
}
```

**Comment by FezVrasta** [comment]
@brentertz would you mind to send a PR to add your last mock somewhere in the docs?

**Comment by oliviertassinari** [comment]
> but now I get NaN for some computed styles

@frontsideair It might explain why I have got some flaky red build on Material-UI.

> Warning: `NaN` is an invalid value for the `left` css style property.

For instance [this one](https://circleci.com/gh/callemall/material-ui/14871?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). I'm using mocha, jest mocking isn't an option for me on this project. 
I have been adding some polyfill to make popper.js testing possible. It sounds like I'm missing something. e.g.:
```js
  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
    },
  });
```

**Comment by brentertz** [comment]
@oliviertassinari I ran into the same things as you. Same polyfill, same NaN, etc.  Even though you are not using Jest, you do have Sinon, which could be used to stub Popper in a similar fashion.

**Comment by illepic** [comment]
A quick modification to avoid having to support ES7 static properties in linters/test-runners:

```js
import PopperJs from 'popper.js';

export default class Popper {
  // static placements = PopperJs.placements;
  constructor() {
    this.placements = PopperJs.placements;

    return {
      destroy: () => {},
      scheduleUpdate: () => {}
    };
  }
}
```

This appears to work for the Popper.js stuff I need to mock in Jest tests.

**Comment by github-account-because-they-want-it** [comment]
I used this from one of the comments above in `node_modules/__mocks__/popper.js.js` but it didn't work:
```
jest.mock('popper.js', () => {
    const PopperJS = jest.requireActual('popper.js');

    return class {
        static placements = PopperJS.placements;

        constructor() {
            return {
                destroy: () => { },
                scheduleUpdate: () => { }
            };
        }
    };
});```
Moving it to jest's setup file worked, though

**Comment by brentertz** [comment]
@github-account-because-they-want-it The aforementioned comment actually said to,  "Create a folder/file called `__mocks__/popper.js.js` **adjacent** to your `node_modules` directory", rather than as a child of `node_modules` as your comment suggests.

Jest is really flexible and provides numerous ways to load mocks, one of which you have demonstrated above.  More can be found here.  https://facebook.github.io/jest/docs/en/manual-mocks.html

**Comment by stream7** [comment]
@FezVrasta I tried both `__mocks__` and setup file but jest seems to pick up the dist file which contains the popper.js already. Any suggestions? I think popper.js should be removed from that file.

```
    unhandledRejection TypeError: document.createRange is not a function
        at findCommonOffsetParent (/Users/niko/Code/example/components/node_modules/react-popper/dist/react-popper.js:765:24)
        at getReferenceOffsets (/Users/niko/Code/example/components/node_modules/react-popper/dist/react-popper.js:1227:28)
        at Popper.update (/Users/niko/Code/example/components/node_modules/react-popper/dist/react-popper.js:1401:28)
        at Popper.update$$1 (/Users/niko/Code/example/components/node_modules/react-popper/dist/react-popper.js:2905:21)
        at /Users/niko/Code/example/components/node_modules/react-popper/dist/react-popper.js:584:7
        at <anonymous>
```

**Comment by FezVrasta** [comment]
Are you using the latest react-popper? Older versions embedded their own Popper.js version

**Comment by FezVrasta** [comment]
Oh fuck, nevermind, it's still like this.. wtf

**Comment by stream7** [comment]
Hmm, I just realized I'm reporting this in the wrong repo, right?

**Comment by FezVrasta** [comment]
Yup, anyway I'm on it.

**Comment by brentertz** [comment]
Trying to update popper to latest version... Mocking it using the above techniques is no longer functional.  Any ideas?

**Comment by FezVrasta** [comment]
I haven't changed anything in the overall structure... Are you sure it is related to the new version?

**Comment by stream7** [comment]
@FezVrasta he is having the same issue I have. I didn't check it thoroughly but I think you need to re-publish the `react-popper` package with the proper dist file which doesn't include popper.

**Comment by FezVrasta** [comment]
`react-popper@next` doesn't include Popper.js, try that. But I don't think that's the problem, Jest shouldn't care.

**Comment by rolandjitsu** [comment]
@FezVrasta I just tried the next version (0.9.0) and the error seems to be gone. But the TS declaration file seems to be gone from the dist.

**Comment by brentertz** [comment]
I tried with `react-popper@0.9.0` and the mock  once again works as expected.  Is the "next" version safe/stable enough to use?

**Comment by FezVrasta** [comment]
The build process has been completely rewritten, there are few bugs, but if it works in your own build process I'd say it's safe to use. Just note that TypeScript annotations are missing for now.

**Comment by jptissot** [comment]
I was having the same issue with enzyme `mount()` and mocking did not work. However,  adding this to my setupTests.js file made everything work. 

``` javascript 

// setup file
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

if (global.document) {
  document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
}
 ```

**Comment by jordonedavidson** [comment]
After following the suggestion of adding the mock file for popper, I receive the following error after every one of my test suites runs:

`console.error node_modules/prop-types/factoryWithTypeCheckers.js:21
Warning: Invalid argument supplied to oneOf, expected an instance of array.`

Ideas?

Im using:
 "@testing-library/jest-dom": "^4.0.0",
 "@testing-library/react": "^8.0.4",

**Comment by tecneider** [comment]
@jordonedavidson https://github.com/atomiks/tippy.js-react/blob/master/test/setup.js

**Comment by stephenh** [comment]
Fwiw/fyi in create-react-app we had to put this at `src/__mocks__/popper.js.js`:

```typescript
import StockPopperJs from "popper.js";

export default class PopperJs {
  static placements = StockPopperJs.placements;

  constructor() {
    return {
      destroy: () => {},
      scheduleUpdate: () => {},
      update: () => {},
    };
  }
}
```

(And, yes, the double-extension `popper.js.js` file name is required.)

**Comment by mentierd** [comment]
Thank you all!  Mocking this significantly reduced the time it took my jest (jsdom) integration tests to execute. anyone who comes across this; I **HIGHLY** recommend you do this. It will not affect the functionality of your tooltip/modal based tests.

**Comment by jashelb90** [comment]
For anyone looking to force the mock to render out in their tests (I was testing using Vue and Jest) I used the following with and additonal render method mock defined..

```js       
        jest.mock('popper.js', () => {
            const PopperJS = jest.requireActual('popper.js');

            return class {
                static placements = PopperJS.placements;

                constructor() {
                    return {
                        destroy: () => { },
                        scheduleUpdate: () => { },
                        render: function (this: any, h){
                            return this.$options._renderChildren
                        }
                    };
                }
            };
        });
```

**Comment by strunkandwhite** [comment]
If anyone's coming to this after upgrading to `@popperjs/core` 2.x.x, this variation on the above worked for me:

```
jest.mock('@popperjs/core', () => {
  const popper = jest.requireActual('@popperjs/core');

  return {
    ...popper,
    createPopper: () => ({
      destroy: () => null,
    }),
  };
});
```

**Comment by jasonwilliams** [comment]
For anyone using jest and react-popper, this worked for me.
Create a file`__mocks__/@popperjs/core.ts` with the contents:

```ts
const mock = () => {
    const PopperJS = jest.requireActual('@popperjs/core');
    return {
        placements: PopperJS.placements,
        destroy: () => {},
        scheduleUpdate: () => {},
        render: function (this: any) {
            return this.$options._renderChildren;
        },
    };
};

export default mock;
export { mock as createPopper };
```

**Comment by Danones** [comment]
Hey 👋🏽 

After reading so many threads, none of the solutions provided seemed to work with me. 

My project currently is using Remix, Happy dom, Vitest and I was using HoverCard from radix.ui. 

I solved the issue by doing the following: 

- Replace `Happy-dom` for `jsdom`. 
- in your test setup file `global.ResizeObserver = require('resize-observer-polyfill');` in order to mock the resize-observer;

This two steps fixed the issue for me. 
Not really sure why I wasn't able to continue using `happy-dom` since they refer they have support for the `resize-observer` but in my tests I kept having the same error as above until I implemented this changes. 

Happy Coding 👋🏽
